freqtrade/optimize/hyperopt.py

```diff
@@ -609,10 +1050,267 @@ def load_previous_results(results_file: Path) -> List:
+            elif type(d).__name__ == 'Real':
+                n_parameters += max(10, int(d.high - d.low))
+            else:
+                n_parameters += len(d.bounds)
+        # guess the size of the search space as the count of the
+        # unordered combination of the dimensions entries
+        try:
+            search_space_size = int(
+                (factorial(n_parameters) /
+                 (factorial(n_parameters - n_dimensions) * factorial(n_dimensions))))
+        except OverflowError:
+            search_space_size = VOID_LOSS
+        # logger.info(f'Search space size: {search_space_size}')
+        log_opt = int(log(opt_points, 2)) if opt_points > 4 else 2
+        if search_space_size < opt_points:
+            # don't waste if the space is small
+            n_initial_points = opt_points // 3
+            min_epochs = opt_points
+        elif total_epochs > 0:
+            # coefficients from total epochs
+            log_epp = int(log(total_epochs, 2)) * log_opt
+            n_initial_points = min(log_epp, total_epochs // 3)
+            min_epochs = total_epochs
+        else:
+            # extract coefficients from the search space
+            log_sss = int(log(search_space_size, 10)) * log_opt
+            # never waste
+            n_initial_points = min(log_sss, search_space_size // 3)
+            # it shall run for this much, I say
+            min_epochs = int(max(n_initial_points, opt_points) + 2 * n_initial_points)
+        return int(n_initial_points * effort) or 1, int(min_epochs * effort), search_space_size
+
+    def update_max_epoch(self, val: Dict, current: int):
+        """ calculate max epochs: store the number of non best epochs
+            between each best, and get the mean of that value """
+        if val['is_initial_point'] is not True:
+            self.epochs_since_last_best.append(current - self.current_best_epoch)
+            self.avg_best_occurrence = (sum(self.epochs_since_last_best) //
+                                        len(self.epochs_since_last_best))
+            self.current_best_epoch = current
+            self.max_epoch = int(
+                (self.current_best_epoch + self.avg_best_occurrence + self.min_epochs) *
+                max(1, self.effort))
+            if self.max_epoch > self.search_space_size:
+                self.max_epoch = self.search_space_size
+        logger.debug(f'\nMax epoch set to: {self.epochs_limit()}')
+
+    @staticmethod
+    def params_Xi(v: dict):
+        return list(v["params_dict"].values())
+
+    def track_points(self, trials: List = None):
+        """
+        keep tracking of the evaluated points per optimizer random state
+        """
+        # if no trials are given, use saved trials
+        if not trials:
+            if len(self.trials) > 0:
+                if self.config.get('hyperopt_continue_filtered', False):
+                    raise ValueError()
+                    # trials = filter_trials(self.trials, self.config)
```

Comment:
the method `filter_trials` is undefined - seems to be missing since it's been introduced (history does not have this method at all).